File: 2016/day-05/day-05.js
const fsp = require("fs/promises")
const path = require("path")
const crypto = require("crypto")

main()

async function main() {
  const input = await parseInput()

  console.log("Part one:", partOne())
  console.log("Part two:", partTwo())

  function partOne(){
    let i = 0
    let password = ""
    while(password.length < 8){
      const toHash = input + i.toString()
      const hash = crypto.createHash("md5").update(toHash).digest("hex")
      if(hash.slice(0, 5) === "00000"){
        password += hash[5]
      }
      i++
    }
  
    return password
  }

  function partTwo(){
    let i = 0
    let password = [null, null, null, null, null, null, null, null]
    while(password.filter(ea => ea !== null).length < 8){
      const toHash = input + i.toString()
      const hash = crypto.createHash("md5").update(toHash).digest("hex")
      if(hash.slice(0, 5) === "00000"){
        if(hash[5] < 8 && password[hash[5]] === null){
          password[hash[5]] = hash[6]
        }
      }
      i++
    }
  
    return password.join("")
  }

  async function parseInput() {
    const rawInput = await fsp.readFile(path.resolve(__dirname, "day-05-input.txt"), "utf-8")
    const input = rawInput.split("\n")
    return input
  }
}

File: 2016/day-05/day-05.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

class Program
{
  static void Main()
  {
    Input input = new Input("day-05-input.txt");
    PartOne partOne = new PartOne(input);
    partOne.Main();
  }
}

class Input
{
  public string DoorId;
  public Input(string filepath)
  {
    DoorId = File.ReadAllText(filepath);
  }
}

class PartOne : PuzzlePart
{
  public PartOne(Input input) : base(input)
  {
  }

  public void Main()
  {
    while(Password.Length < 8)
    {
      MD5 hasher = MD5.Create();
      byte[] convertedInput = Encoding.ASCII.GetBytes($"{_input}{Index}");
      byte[] hash = hasher.ComputeHash(convertedInput);
      string hexString = String.Join("", BitConverter.ToString(hash).Split("-"));
      if(hexString.Substring(0,5) == "00000")
      {
        Password += hexString[5];
      }
      Index++;
    }
    System.Console.WriteLine($"Password: {Password}");
  }
}

class PartTwo : PuzzlePart
{
  public PartTwo(Input input) : base(input)
  {
  }
  
  public void Main()
  {

  }
}

class PuzzlePart
{

  protected string _input;
  public string Password { get; set; }
  public int Index { get; set; }
  public PuzzlePart(Input input)
  {
    _input = input.DoorId;
    // _input = "abc";
    Index = 0;
    Password = "";
  }
}